plugins/modem-manager/fu-mm-utils.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2019 Aleksander Morgado <aleksander@aleksander.es>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include <gio/gio.h>
+#include "fu-mm-utils.h"
+
+gboolean
+fu_mm_utils_get_udev_port_info (GUdevDevice  *device,
+                                gchar       **out_device_sysfs_path,
+                                gint         *out_port_ifnum,
+                                GError      **error)
+{
+        g_autoptr(GUdevDevice) parent = NULL;
+        g_autofree gchar *device_sysfs_path = NULL;
+        gint port_ifnum = -1;
+        const gchar *aux;
+
+        /* ID_USB_INTERFACE_NUM is set on the port device itself */
+        aux = g_udev_device_get_property (device, "ID_USB_INTERFACE_NUM");
+        if (aux != NULL)
+                port_ifnum = (guint16) g_ascii_strtoull (aux, NULL, 16);
+
+        /* we need to traverse all parents of the give udev device until we find
+         * the first 'usb_device' reported, which is the GUdevDevice associated with
+         * the full USB device (i.e. all ports of the same device).
+         */
+        parent = g_udev_device_get_parent (device);
+        while (parent != NULL) {
+                GUdevDevice *next;
+
+                if (g_strcmp0 (g_udev_device_get_devtype (parent), "usb_device") == 0) {
+                        device_sysfs_path = g_strdup (g_udev_device_get_sysfs_path (parent));
+                        break;
+                }
+
+                /* Check next parent */
+                next = g_udev_device_get_parent (parent);
+                g_object_unref (parent);
+                parent = next;
```

Comment:
`g_set_object (&parent, next)`?

Reply:
doesn't that also increase the refcount of next? although then I could make next a g_autoptr, right?

Reply:
done in a new fixup commit